l3backend/l3backend-color.dtx

```diff
@@ -1412,19 +1413,32 @@ end
 %   The mode is hard-coded as $2$: a PDF literal.
 %    \begin{macrocode}
 local function set_color(head, n, new_color, old_color)
-  if new_color == old_color or not new_color then return head end
-  local color_node = node_new("whatsit","pdf_literal")
-  color_node.mode = 2
-  color_node.data = new_color
+  if new_color == old_color then return head end
+  local color_node = node_new("whatsit","pdf_colorstack")
```

Comment:
Perhaps I'm missing something, but how does this select the appropriate colour stack?

Reply:
Colors are controlled by colorstack 0 which is the default after `node_new`. We should probably make this more explicit.

Reply:
That's going to need some thought. How does this work in a case where you'd need a separate colour stack in pdfTeX, e.g. footnotes breaking over pages? Are we 'OK' as we are dealing with nodes?

Reply:
As long as the user is actually using `l3color` it's fine since the attribute approach handles footnotes correctly anyway.

But using non attribute colors based on another colorstack interleaved with attribute based colors might reset the wrong stack. We can avoid that by storing the stack number as an additional attribute or by adding a user defined whatsit in order to switch the currently used colorstack. (Which make more sense depends a bit on where the nodes are created. E.g. if a saved box is added in the middle of a footnote it should probably use the footnote stack but that might not be known when the box is created... Then again using classical color commands there would probably use the main colorstack, so using them for luacolors too would be more consistent.)

In any case this is rather easy to adapt later if the issue ever actually appears.

Reply:
OK, that's fine then: I think we look hard at getting all colour in LuaTeX to the attribute-based approach.